From a43fa3202c213437e75e82986bc9f1d749b26661 Mon Sep 17 00:00:00 2001
From: Richard Guy Briggs <rgb@redhat.com>
Date: Tue, 23 Sep 2014 03:59:06 -0400
Subject: [kernel] audit: convert PPIDs to the inital PID namespace

Message-id: <9e9499d072513565693504791c3210be94b961c0.1411415193.git.rgb@redhat.com>
Patchwork-id: 96454
O-Subject: [RHEL7 PATCH 3/5] audit: convert PPIDs to the inital PID namespace.
Bugzilla: 1010455
RH-Acked-by: Paul Moore <pmoore@redhat.com>

sys_getppid() returns the parent pid of the current process in its own pid
namespace.  Since audit filters are based in the init pid namespace, a process
could avoid a filter or trigger an unintended one by being in an alternate pid
namespace or log meaningless information.

Switch to task_ppid_nr() for PPIDs to anchor all audit filters in the
init_pid_ns.

Upstream: c92cdeb
(informed by ebiederman's 6c621b7e)
Cc: stable@vger.kernel.org
Cc: Eric W. Biederman <ebiederm@xmission.com>
Signed-off-by: Richard Guy Briggs <rgb@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/audit.c b/kernel/audit.c
index 6129138..93501a8 100644
--- a/kernel/audit.c
+++ b/kernel/audit.c
@@ -1739,10 +1739,10 @@ void audit_log_task_info(struct audit_buffer *ab, struct task_struct *tsk)
  spin_unlock_irq(&tsk->sighand->siglock);
 
  audit_log_format(ab,
-    " ppid=%ld pid=%d auid=%u uid=%u gid=%u"
+    " ppid=%d pid=%d auid=%u uid=%u gid=%u"
     " euid=%u suid=%u fsuid=%u"
     " egid=%u sgid=%u fsgid=%u tty=%s ses=%u",
-    sys_getppid(),
+    task_ppid_nr(tsk),
     tsk->pid,
     from_kuid(&init_user_ns, audit_get_loginuid(tsk)),
     from_kuid(&init_user_ns, cred->uid),
diff --git a/kernel/auditsc.c b/kernel/auditsc.c
index 989c24e..59ed726 100644
--- a/kernel/auditsc.c
+++ b/kernel/auditsc.c
@@ -459,7 +459,7 @@ static int audit_filter_rules(struct task_struct *tsk,
   case AUDIT_PPID:
    if (ctx) {
     if (!ctx->ppid)
-     ctx->ppid = sys_getppid();
+     ctx->ppid = task_ppid_nr(tsk);
     result = audit_comparator(ctx->ppid, f->op, f->val);
    }
    break;
-- 
1.7.1